src/clj/rems/auth/oidc.clj

```diff
@@ -23,25 +26,71 @@
-    (some #(get id-data (keyword %)) attrs)))
+(defn- get-userid-attributes [id-data]
+  (for [{:keys [attribute rename]} (getx env :oidc-userid-attributes)
+        :let [value (get id-data (keyword attribute))]
+        :when value]
+    [(keyword (or rename attribute)) value]))
+
+(deftest test-get-userid-attributes
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "sub" :rename "elixirId"}
+                                              {:attribute "old_sub"}]}]
+    (is (= [] (get-userid-attributes nil)))
+    (is (= [[:elixirId "elixir-alice"]
+            [:old_sub "alice"]]
+           (get-userid-attributes {:old_sub "alice"
+                                   :sub "elixir-alice"
+                                   :name "Alice Applicant"})))))
+
+(defn get-userid [id-data]
+  (second (first (get-userid-attributes id-data))))
 
 (deftest test-get-userid
-  (with-redefs [env {:oidc-userid-attribute "atr"}]
-    (is (nil? (get-userid {:sub "123"})))
-    (is (= "456" (get-userid {:sub "123" :atr "456"}))))
-  (with-redefs [env {:oidc-userid-attribute ["atr"]}]
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr"}]}]
     (is (nil? (get-userid {:sub "123"})))
     (is (= "456" (get-userid {:sub "123" :atr "456"}))))
-  (with-redefs [env {:oidc-userid-attribute ["atr" "sub" "fallback"]}]
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr"}
+                                              {:attribute "sub"}
+                                              {:attribute "fallback"}]}]
     (is (= "123" (get-userid {:sub "123"})))
     (is (= "456" (get-userid {:sub "123" :atr "456"})))
     (is (= "456" (get-userid {:sub "123" :atr "456" :fallback "78"})))
-    (is (= "78" (get-userid {:fallback "78"})))))
+    (is (= "78" (get-userid {:fallback "78"}))))
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr" :rename "elixirId"}
+                                              {:attribute "sub"}]}]
+    (is (= "elixir-alice" (get-userid {:atr "elixir-alice" :sub "123"})))
+    (is (= "123" (get-userid {:elixirId "elixir-alice" :sub "123"}))
+        "user data should use names before rename")))
+
+(defn find-or-create-user! [id-data]
+  (let [userid-attrs (get-userid-attributes id-data)]
+    (or (some #(user-mappings/get-user-mapping (first %) (second %)) userid-attrs)
+        (find-first users/user-exists? (map second userid-attrs))
+        (let [userid (get-userid id-data)]
+          (users/add-user-raw! userid id-data)
+          userid))))
+
+(defn save-user-mappings! [id-data userid]
+  (let [attrs (get-userid-attributes id-data)]
+    (doseq [[attr value] attrs
+            :when (not= value userid)]
+      (user-mappings/create-user-mapping! {:userid userid
+                                           :ext-id-attribute (name attr)
+                                           :ext-id-value value}))))
+
+(defn get-user-attributes [id-data]
+  (let [identity-base {:eppn (get-userid id-data)
```

Comment:
`:eppn` seems a little redundant here (looking at fake-login usage) but we can refactor it with #2771 if needed

Reply:
Or here https://github.com/CSCfi/rems/issues/2377